File: codittle_db.py
#!/usr/bin/env python3
"""
codittle_db.py -- inspect / back up / RECOVER Codittle's embedded PGlite database.

Codittle stores SSH connections (and projects) in a per-workspace PGlite database at
  %LOCALAPPDATA%\\Codittle\\workspaces\\<id>\\pgdata
If that DB corrupts, Codittle quarantines it as `pgdata.corrupt-<ts>` and starts a
FRESH EMPTY one -> your connections vanish from the UI even though the data still
exists in another workspace (or a backup). This tool finds it and puts it back.

  python codittle_db.py status                          # workspaces, the active one,
                                                         #   connection counts, corruption flags
  python codittle_db.py connections [--workspace ID]     # list connections in a workspace
  python codittle_db.py backup [--all]                   # snapshot pgdata -> DevTools/_codittle-db-backups/
  python codittle_db.py restore [--from ID|PATH] [--apply]
        # copy connections INTO the active workspace, re-pointed to its stream.
        # Auto-source = the other workspace with the most connections.

SAFE BY DEFAULT:
  * restore is a DRY RUN unless --apply
  * --apply REFUSES to run while Codittle is open (writing a live DB is what corrupts it)
  * --apply always backs up the active DB first
Requires Codittle installed (uses its bundled node.exe for PGlite); read-only on sources.
"""
import argparse
import json
import os
import shutil
import subprocess
import sys
from datetime import datetime
from pathlib import Path

import codittle_connections as cc   # reuse the Codittle node.exe finder

WS = Path(os.environ.get("LOCALAPPDATA", "")) / "Codittle" / "workspaces"
MJS = Path(__file__).resolve().parent / "codittle_db.mjs"
BACKUPS = Path(__file__).resolve().parents[1] / "_codittle-db-backups"   # DevTools/_codittle-db-backups


def _node():
    n = cc._find_codittle_node()
    if not n:
        sys.exit("ERROR: Codittle node.exe not found (install/run Codittle once).")
    return str(n)


def _mjs(*args):
    out = subprocess.run([_node(), str(MJS), *args], capture_output=True, text=True)
    if out.returncode != 0:
        sys.exit(f"db query failed: {out.stderr.strip() or out.stdout.strip()}")
    return json.loads(out.stdout or "null")


def _codittle_running():
    try:
        out = subprocess.run(["tasklist"], capture_output=True, text=True, timeout=5).stdout.lower()
        return "codittle-desktop.exe" in out
    except Exception:
        return False


def _status():
    return _mjs("status")


def _active(st):
    return next((w for w in st if w.get("active")), st[0] if st else None)


def _best_source(st):
    cands = [w for w in st if not w.get("active") and (w["connections"] or 0) > 0]
    return max(cands, key=lambda w: w["connections"], default=None)


def _resolve_pgdata(ref):
    p = Path(ref)
    if p.name == "pgdata" and p.is_dir():
        return str(p)
    if (p / "pgdata").is_dir():
        return str(p / "pgdata")
    if (WS / ref / "pgdata").is_dir():
        return str(WS / ref / "pgdata")
    if p.is_dir():
        subs = list(p.glob("*-pgdata")) or list(p.glob("*/pgdata"))
        if subs:
            return str(subs[0])
    sys.exit(f"could not resolve a pgdata directory from '{ref}'")


# ── commands ────────────────────────────────────────────────────────────────
def cmd_status(_):
    st = _status()
    print(f"workspaces: {WS}\n")
    for w in st:
        tag = "ACTIVE" if w.get("active") else "  --  "
        conn = w["connections"] if w["connections"] >= 0 else "UNREADABLE(corrupt)"
        corr = f"  corrupt-snapshots={len(w['corrupt_snapshots'])}" if w["corrupt_snapshots"] else ""
        print(f"[{tag}] {w['workspace']}  connections={conn}  projects={w['projects']}{corr}")
    a = _active(st)
    if a and (a["connections"] or 0) == 0:
        src = _best_source(st)
        if src:
            print(f"\n!  active workspace has 0 connections — {src['workspace']} has {src['connections']}.")
            print("   recover (close Codittle first):")
            print(f"     python codittle_db.py restore --from {src['workspace']} --apply")


def cmd_connections(args):
    st = _status()
    w = next((x for x in st if x["workspace"] == args.workspace), None) if args.workspace else _active(st)
    if not w:
        sys.exit("workspace not found")
    rows = _mjs("connections", w["pgdata"])
    for r in rows:
        print(f"  {r['name']:12} {r['username']}@{r['host']}:{r['port']}  {r.get('ssh_home') or ''}")
    print(f"\n{len(rows)} connections in {w['workspace']}")


def cmd_backup(args):
    st = _status()
    dest = BACKUPS / f"{datetime.now():%Y%m%d-%H%M%S}-manual"
    for w in (st if args.all else [_active(st)]):
        d = dest / f"{w['workspace']}-pgdata"
        shutil.copytree(w["pgdata"], d)
        print(f"backed up {w['workspace']} -> {d}")


def cmd_restore(args):
    st = _status()
    dst = _active(st)
    if not dst:
        sys.exit("no active workspace found")
    src = _resolve_pgdata(args.from_) if args.from_ else (_best_source(st) or {}).get("pgdata")
    if not src:
        sys.exit("no other workspace has connections — pass --from <id|backup-path>")
    print(f"source      : {src}")
    print(f"target(active): {dst['pgdata']}")

    if not args.apply:
        res = _mjs("restore", src, dst["pgdata"], "dry")
        print(f"\nDRY RUN — would restore {res['would_restore']}: {', '.join(res['names'])}")
        print("Close Codittle, then re-run with --apply to write.")
        return

    if _codittle_running() and not args.force:
        sys.exit("Codittle is RUNNING — close it first (writing a live DB corrupts it). "
                 "Quit Codittle fully, then re-run. (--force to override, not advised.)")

    bk = BACKUPS / f"{datetime.now():%Y%m%d-%H%M%S}-preapply" / f"{dst['workspace']}-pgdata"
    shutil.copytree(dst["pgdata"], bk)
    print(f"backup      : {bk}")
    res = _mjs("restore", src, dst["pgdata"], "apply")
    print(f"\nrestored {res['added']} connection(s): {', '.join(res['names'])}")
    print(f"active workspace now has {res['total']}. Reopen Codittle to see them.")
    print("(passwords are copied encrypted — if any fails to auth, re-enter it; they're in Test_Environments.csv)")


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    sub = ap.add_subparsers(dest="cmd", required=True)
    sub.add_parser("status")
    c = sub.add_parser("connections"); c.add_argument("--workspace")
    b = sub.add_parser("backup"); b.add_argument("--all", action="store_true")
    r = sub.add_parser("restore")
    r.add_argument("--from", dest="from_", metavar="ID|PATH", help="source workspace id or backup path")
    r.add_argument("--apply", action="store_true", help="actually write (default is dry run)")
    r.add_argument("--force", action="store_true", help="write even if Codittle seems to be running")
    args = ap.parse_args()
    {"status": cmd_status, "connections": cmd_connections,
     "backup": cmd_backup, "restore": cmd_restore}[args.cmd](args)


if __name__ == "__main__":
    main()

File: codittle_db.mjs
/**
 * codittle_db.mjs — node side of codittle_db.py (run with Codittle's own node.exe).
 *
 *   node codittle_db.mjs status
 *   node codittle_db.mjs connections <pgdata>
 *   node codittle_db.mjs restore <srcPgdata> <dstPgdata> <dry|apply>
 *
 * Reads are done on a temp COPY (never touch the source / a live DB). The restore
 * 'apply' path writes directly into <dstPgdata> — the caller (codittle_db.py) must
 * ensure Codittle is closed first.
 */
import { cpSync, rmSync, existsSync, readdirSync, statSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';

const execBase  = 'file:///' + process.execPath.replace(/\\/g, '/');
const pgliteUrl = new URL('./sidecar/node_modules/@electric-sql/pglite/dist/index.js', execBase).href;
const { PGlite } = await import(pgliteUrl);

const WORKSPACES = join(process.env['LOCALAPPDATA'] ?? '', 'Codittle', 'workspaces');

function openCopy(pgdata) {
  const copyDir = join(tmpdir(), 'cdb-' + Date.now() + '-' + Math.random().toString(36).slice(2, 7));
  cpSync(pgdata, copyDir, { recursive: true });
  const pid = join(copyDir, 'postmaster.pid'); if (existsSync(pid)) rmSync(pid);
  return { db: new PGlite('file://' + copyDir.replace(/\\/g, '/')), copyDir };
}

async function countWs(pgdata) {
  const { db, copyDir } = openCopy(pgdata);
  let envs = -1, projects = -1;
  try {
    envs = (await db.query('SELECT count(*)::int n FROM t24_stream_envs')).rows[0].n;
    projects = (await db.query('SELECT count(*)::int n FROM projects')).rows[0].n;
  } catch { /* corrupt / unreadable */ }
  await db.close(); rmSync(copyDir, { recursive: true, force: true });
  return { envs, projects };
}

const action = process.argv[2];

if (action === 'status') {
  const out = [];
  for (const name of (existsSync(WORKSPACES) ? readdirSync(WORKSPACES) : [])) {
    const pg = join(WORKSPACES, name, 'pgdata');
    if (!existsSync(pg)) continue;
    const corrupt = readdirSync(join(WORKSPACES, name)).filter(n => n.startsWith('pgdata.corrupt'));
    const { envs, projects } = await countWs(pg);
    out.push({ workspace: name, pgdata: pg, mtime: statSync(pg).mtimeMs,
               connections: envs, projects, corrupt_snapshots: corrupt });
  }
  out.sort((a, b) => b.mtime - a.mtime);
  if (out.length) out[0].active = true;          // Codittle uses the most-recent pgdata
  console.log(JSON.stringify(out, null, 2));

} else if (action === 'connections') {
  const { db, copyDir } = openCopy(process.argv[3]);
  const rows = (await db.query('SELECT name, host, port, username, ssh_home FROM t24_stream_envs ORDER BY created_at')).rows;
  await db.close(); rmSync(copyDir, { recursive: true, force: true });
  console.log(JSON.stringify(rows, null, 2));

} else if (action === 'restore') {
  const [src, dst, mode] = process.argv.slice(3);
  const { db: srcDb, copyDir } = openCopy(src);
  const envs = (await srcDb.query('SELECT * FROM t24_stream_envs ORDER BY created_at')).rows;
  await srcDb.close(); rmSync(copyDir, { recursive: true, force: true });

  if (mode !== 'apply') {
    console.log(JSON.stringify({ dry_run: true, would_restore: envs.length, names: envs.map(e => e.name) }, null, 2));
    process.exit(0);
  }
  const dpid = join(dst, 'postmaster.pid'); if (existsSync(dpid)) rmSync(dpid);
  const dstDb = new PGlite('file://' + dst.replace(/\\/g, '/'));
  const stream = (await dstDb.query('SELECT id FROM t24_release_streams ORDER BY created_at LIMIT 1')).rows[0];
  if (!stream) { console.error('no release stream in destination'); process.exit(1); }
  const existing = (await dstDb.query('SELECT host FROM t24_stream_envs')).rows.map(r => r.host);
  let added = 0; const names = [];
  for (const e of envs) {
    if (existing.includes(e.host)) continue;     // don't duplicate
    const id = 'env_' + Date.now().toString(36) + Math.random().toString(36).slice(2, 8);
    const now = new Date().toISOString();
    await dstDb.query(
      `INSERT INTO t24_stream_envs
        (id, stream_id, env_class, host, port, username, password_enc, remote_roots, ssh_home, exec_setup, name, last_viewed_at, created_at, updated_at)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14)`,
      [id, stream.id, e.env_class, e.host, e.port, e.username, e.password_enc ?? null,
       e.remote_roots ?? null, e.ssh_home ?? null, e.exec_setup ?? null, e.name, null, now, now]);
    added++; names.push(e.name);
  }
  const total = (await dstDb.query('SELECT count(*)::int n FROM t24_stream_envs')).rows[0].n;
  await dstDb.close();
  if (existsSync(dpid)) rmSync(dpid);            // leave it unlocked for Codittle
  console.log(JSON.stringify({ applied: true, added, total, names }, null, 2));

} else {
  console.error('usage: status | connections <pgdata> | restore <src> <dst> <dry|apply>');
  process.exit(2);
}
